<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from robot_properties_fingers/xacro/pro/trifingerpro.urdf.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="trifingerpro">
  <!--
    Frames are defined such that they all align with the base frame when all
    joints are at position zero.

    In zero-configuration the origins of all joints are in one line.

    In zero-configuration the origin of the base link is exactly above the
    finger tip which should make it easy to place the finger in the world.
    -->
  <material name="fingerpro_shell">
    <!-- colour from image -->
    <color rgba="0.804 0.764 0.761 1.0"/>
  </material>
  <material name="fingerpro_tip">
    <!-- colour from image -->
    <color rgba="0.33 0.36 0.37 1.0"/>
  </material>
  <material name="fingerpro_motor">
    <color rgba="0.17 0.17 0.17 1.0"/>
  </material>
  <material name="center_holder_material">
    <color rgba="0.1 0.1 0.1 1.0"/>
  </material>
  <!-- Define the global base_link and place all other objects relative to it. -->
  <link name="base_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="4.166666666666667e-06" ixy="0" ixz="0" iyy="4.166666666666667e-06" iyz="0" izz="4.166666666666667e-06"/>
    </inertial>
  </link>
  <!-- Define the upper holder -->
  <link name="upper_holder_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Center.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="center_holder_material"/>
    </visual>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="4.166666666666667e-06" ixy="0" ixz="0" iyy="4.166666666666667e-06" iyz="0" izz="4.166666666666667e-06"/>
    </inertial>
  </link>
  <joint name="base_to_upper_holder_joint" type="fixed">
    <parent link="base_link"/>
    <child link="upper_holder_link"/>
    <origin xyz="0 0 0.29"/>
  </joint>
  <!--
        Fixed links for the finger base (parts where the upper link is mounted).
        Note: Link Frames already coincide with the joint location
        -->
  <link name="finger_base_link_0">
    <visual>
      <origin rpy="0 0 0" xyz="0 0.23295 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Base.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0.23295 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Base.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="0.00041666666666666675" ixy="0" ixz="0" iyy="0.00041666666666666675" iyz="0" izz="0.00041666666666666675"/>
    </inertial>
  </link>
  <link name="finger_upper_link_0">
    <visual>
      <origin rpy="0 0 0" xyz="0 0.0235 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Proximal.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0.0235 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Proximal.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0.06 0"/>
      <mass value="0.26"/>
      <inertia ixx="0.0004593333333333334" ixy="0" ixz="0" iyy="6.933333333333334e-05" iyz="0" izz="0.0004593333333333334"/>
    </inertial>
  </link>
  <link name="finger_upper_link_motor_0">
    <!-- motor -->
    <visual>
      <origin xyz="0.018 0.141 -0.001"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Motor.stl"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- end stop -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="0.015 0.06 0"/>
      <geometry>
        <cylinder length="0.015" radius="0.005"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- cables -->
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="-0.008 0.1 0"/>
      <geometry>
        <cylinder length="0.12" radius="0.002"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- opening on lower joint -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="-0.008 0.026 0"/>
      <geometry>
        <cylinder length="0.002" radius="0.014"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
  </link>
  <joint name="finger_upper_to_motor_joint_0" type="fixed">
    <parent link="finger_upper_link_0"/>
    <child link="finger_upper_link_motor_0"/>
    <origin xyz="0 0.0235 0"/>
  </joint>
  <link name="finger_middle_link_0">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Intermediate.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Intermediate.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.028 0 -0.08"/>
      <mass value="0.25"/>
      <inertia ixx="0.0004416666666666667" ixy="0" ixz="0" iyy="0.0004416666666666667" iyz="0" izz="6.666666666666667e-05"/>
    </inertial>
  </link>
  <link name="finger_middle_link_motor_0">
    <!-- motor -->
    <visual>
      <origin xyz="0.048 0 -0.046"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Motor.stl"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- end stop -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="0.048 0 -0.13"/>
      <geometry>
        <cylinder length="0.015" radius="0.005"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- cables -->
    <visual>
      <origin xyz="0.025 0 -0.09"/>
      <geometry>
        <cylinder length="0.12" radius="0.002"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- opening on lower joint -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="0.024 0 -0.16"/>
      <geometry>
        <cylinder length="0.002" radius="0.014"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
  </link>
  <joint name="finger_middle_to_motor_joint_0" type="fixed">
    <parent link="finger_middle_link_0"/>
    <child link="finger_middle_link_motor_0"/>
  </joint>
  <link name="finger_lower_link_0">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_without_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_without_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 -0.06"/>
      <mass value="0.021"/>
      <inertia ixx="3.500000000000001e-05" ixy="0" ixz="0" iyy="3.500000000000001e-05" iyz="0" izz="1.4000000000000001e-06"/>
    </inertial>
  </link>
  <!-- fixed link for finger tip -->
  <link name="finger_tip_link_0">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="-0.0185 0 0.1626"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_actual_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_tip"/>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="-0.0185 0 0.1626"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_actual_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.031"/>
      <inertia ixx="5.166666666666667e-07" ixy="0" ixz="0" iyy="5.166666666666667e-07" iyz="0" izz="5.166666666666667e-07"/>
    </inertial>
  </link>
  <joint name="finger_lower_to_tip_joint_0" type="fixed">
    <parent link="finger_lower_link_0"/>
    <child link="finger_tip_link_0"/>
    <origin xyz="0.0185 0 -0.1626"/>
  </joint>
  <!-- kinematics -->
  <!-- NOTE: Joint limits are set according to soft limits in the
             configuration of the real robot (i.e. not according to physical
             limits). -->
  <joint name="finger_base_to_upper_joint_0" type="revolute">
    <parent link="finger_base_link_0"/>
    <child link="finger_upper_link_0"/>
    <limit effort="1000" lower="-0.33" upper="1.0" velocity="1000"/>
    <axis xyz="0 1 0"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <joint name="finger_upper_to_middle_joint_0" type="revolute">
    <parent link="finger_upper_link_0"/>
    <child link="finger_middle_link_0"/>
    <limit effort="1000" lower="0.0" upper="1.57" velocity="1000"/>
    <axis xyz="1 0 0"/>
    <origin rpy="0 0 0" xyz="0.01685 0.0505 0"/>
  </joint>
  <joint name="finger_middle_to_lower_joint_0" type="revolute">
    <parent link="finger_middle_link_0"/>
    <child link="finger_lower_link_0"/>
    <limit effort="1000" lower="-2.7" upper="0.0" velocity="1000"/>
    <axis xyz="1 0 0"/>
    <origin rpy="0 0 0" xyz="0.04922 0 -0.16"/>
  </joint>
  <!--
        Fixed links for the finger base (parts where the upper link is mounted).
        Note: Link Frames already coincide with the joint location
        -->
  <link name="finger_base_link_120">
    <visual>
      <origin rpy="0 0 0" xyz="0 0.23295 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Base.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0.23295 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Base.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="0.00041666666666666675" ixy="0" ixz="0" iyy="0.00041666666666666675" iyz="0" izz="0.00041666666666666675"/>
    </inertial>
  </link>
  <link name="finger_upper_link_120">
    <visual>
      <origin rpy="0 0 0" xyz="0 0.0235 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Proximal.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0.0235 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Proximal.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0.06 0"/>
      <mass value="0.26"/>
      <inertia ixx="0.0004593333333333334" ixy="0" ixz="0" iyy="6.933333333333334e-05" iyz="0" izz="0.0004593333333333334"/>
    </inertial>
  </link>
  <link name="finger_upper_link_motor_120">
    <!-- motor -->
    <visual>
      <origin xyz="0.018 0.141 -0.001"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Motor.stl"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- end stop -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="0.015 0.06 0"/>
      <geometry>
        <cylinder length="0.015" radius="0.005"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- cables -->
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="-0.008 0.1 0"/>
      <geometry>
        <cylinder length="0.12" radius="0.002"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- opening on lower joint -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="-0.008 0.026 0"/>
      <geometry>
        <cylinder length="0.002" radius="0.014"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
  </link>
  <joint name="finger_upper_to_motor_joint_120" type="fixed">
    <parent link="finger_upper_link_120"/>
    <child link="finger_upper_link_motor_120"/>
    <origin xyz="0 0.0235 0"/>
  </joint>
  <link name="finger_middle_link_120">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Intermediate.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Intermediate.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.028 0 -0.08"/>
      <mass value="0.25"/>
      <inertia ixx="0.0004416666666666667" ixy="0" ixz="0" iyy="0.0004416666666666667" iyz="0" izz="6.666666666666667e-05"/>
    </inertial>
  </link>
  <link name="finger_middle_link_motor_120">
    <!-- motor -->
    <visual>
      <origin xyz="0.048 0 -0.046"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Motor.stl"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- end stop -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="0.048 0 -0.13"/>
      <geometry>
        <cylinder length="0.015" radius="0.005"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- cables -->
    <visual>
      <origin xyz="0.025 0 -0.09"/>
      <geometry>
        <cylinder length="0.12" radius="0.002"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- opening on lower joint -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="0.024 0 -0.16"/>
      <geometry>
        <cylinder length="0.002" radius="0.014"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
  </link>
  <joint name="finger_middle_to_motor_joint_120" type="fixed">
    <parent link="finger_middle_link_120"/>
    <child link="finger_middle_link_motor_120"/>
  </joint>
  <link name="finger_lower_link_120">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_without_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_without_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 -0.06"/>
      <mass value="0.021"/>
      <inertia ixx="3.500000000000001e-05" ixy="0" ixz="0" iyy="3.500000000000001e-05" iyz="0" izz="1.4000000000000001e-06"/>
    </inertial>
  </link>
  <!-- fixed link for finger tip -->
  <link name="finger_tip_link_120">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="-0.0185 0 0.1626"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_actual_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_tip"/>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="-0.0185 0 0.1626"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_actual_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.031"/>
      <inertia ixx="5.166666666666667e-07" ixy="0" ixz="0" iyy="5.166666666666667e-07" iyz="0" izz="5.166666666666667e-07"/>
    </inertial>
  </link>
  <joint name="finger_lower_to_tip_joint_120" type="fixed">
    <parent link="finger_lower_link_120"/>
    <child link="finger_tip_link_120"/>
    <origin xyz="0.0185 0 -0.1626"/>
  </joint>
  <!-- kinematics -->
  <!-- NOTE: Joint limits are set according to soft limits in the
             configuration of the real robot (i.e. not according to physical
             limits). -->
  <joint name="finger_base_to_upper_joint_120" type="revolute">
    <parent link="finger_base_link_120"/>
    <child link="finger_upper_link_120"/>
    <limit effort="1000" lower="-0.33" upper="1.0" velocity="1000"/>
    <axis xyz="0 1 0"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <joint name="finger_upper_to_middle_joint_120" type="revolute">
    <parent link="finger_upper_link_120"/>
    <child link="finger_middle_link_120"/>
    <limit effort="1000" lower="0.0" upper="1.57" velocity="1000"/>
    <axis xyz="1 0 0"/>
    <origin rpy="0 0 0" xyz="0.01685 0.0505 0"/>
  </joint>
  <joint name="finger_middle_to_lower_joint_120" type="revolute">
    <parent link="finger_middle_link_120"/>
    <child link="finger_lower_link_120"/>
    <limit effort="1000" lower="-2.7" upper="0.0" velocity="1000"/>
    <axis xyz="1 0 0"/>
    <origin rpy="0 0 0" xyz="0.04922 0 -0.16"/>
  </joint>
  <!--
        Fixed links for the finger base (parts where the upper link is mounted).
        Note: Link Frames already coincide with the joint location
        -->
  <link name="finger_base_link_240">
    <visual>
      <origin rpy="0 0 0" xyz="0 0.23295 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Base.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0.23295 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Base.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="0.00041666666666666675" ixy="0" ixz="0" iyy="0.00041666666666666675" iyz="0" izz="0.00041666666666666675"/>
    </inertial>
  </link>
  <link name="finger_upper_link_240">
    <visual>
      <origin rpy="0 0 0" xyz="0 0.0235 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Proximal.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0.0235 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Proximal.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0.06 0"/>
      <mass value="0.26"/>
      <inertia ixx="0.0004593333333333334" ixy="0" ixz="0" iyy="6.933333333333334e-05" iyz="0" izz="0.0004593333333333334"/>
    </inertial>
  </link>
  <link name="finger_upper_link_motor_240">
    <!-- motor -->
    <visual>
      <origin xyz="0.018 0.141 -0.001"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Motor.stl"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- end stop -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="0.015 0.06 0"/>
      <geometry>
        <cylinder length="0.015" radius="0.005"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- cables -->
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="-0.008 0.1 0"/>
      <geometry>
        <cylinder length="0.12" radius="0.002"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- opening on lower joint -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="-0.008 0.026 0"/>
      <geometry>
        <cylinder length="0.002" radius="0.014"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
  </link>
  <joint name="finger_upper_to_motor_joint_240" type="fixed">
    <parent link="finger_upper_link_240"/>
    <child link="finger_upper_link_motor_240"/>
    <origin xyz="0 0.0235 0"/>
  </joint>
  <link name="finger_middle_link_240">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Intermediate.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Intermediate.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.028 0 -0.08"/>
      <mass value="0.25"/>
      <inertia ixx="0.0004416666666666667" ixy="0" ixz="0" iyy="0.0004416666666666667" iyz="0" izz="6.666666666666667e-05"/>
    </inertial>
  </link>
  <link name="finger_middle_link_motor_240">
    <!-- motor -->
    <visual>
      <origin xyz="0.048 0 -0.046"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Motor.stl"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- end stop -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="0.048 0 -0.13"/>
      <geometry>
        <cylinder length="0.015" radius="0.005"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- cables -->
    <visual>
      <origin xyz="0.025 0 -0.09"/>
      <geometry>
        <cylinder length="0.12" radius="0.002"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
    <!-- opening on lower joint -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="0.024 0 -0.16"/>
      <geometry>
        <cylinder length="0.002" radius="0.014"/>
      </geometry>
      <material name="fingerpro_motor"/>
    </visual>
  </link>
  <joint name="finger_middle_to_motor_joint_240" type="fixed">
    <parent link="finger_middle_link_240"/>
    <child link="finger_middle_link_motor_240"/>
  </joint>
  <link name="finger_lower_link_240">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_without_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_shell"/>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_without_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 -0.06"/>
      <mass value="0.021"/>
      <inertia ixx="3.500000000000001e-05" ixy="0" ixz="0" iyy="3.500000000000001e-05" iyz="0" izz="1.4000000000000001e-06"/>
    </inertial>
  </link>
  <!-- fixed link for finger tip -->
  <link name="finger_tip_link_240">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="-0.0185 0 0.1626"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_actual_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
      <material name="fingerpro_tip"/>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="-0.0185 0 0.1626"/>
      <geometry>
        <mesh filename="package://robot_properties_fingers/meshes/pro/SIM__BL-Finger_Tip_actual_tip.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.031"/>
      <inertia ixx="5.166666666666667e-07" ixy="0" ixz="0" iyy="5.166666666666667e-07" iyz="0" izz="5.166666666666667e-07"/>
    </inertial>
  </link>
  <joint name="finger_lower_to_tip_joint_240" type="fixed">
    <parent link="finger_lower_link_240"/>
    <child link="finger_tip_link_240"/>
    <origin xyz="0.0185 0 -0.1626"/>
  </joint>
  <!-- kinematics -->
  <!-- NOTE: Joint limits are set according to soft limits in the
             configuration of the real robot (i.e. not according to physical
             limits). -->
  <joint name="finger_base_to_upper_joint_240" type="revolute">
    <parent link="finger_base_link_240"/>
    <child link="finger_upper_link_240"/>
    <limit effort="1000" lower="-0.33" upper="1.0" velocity="1000"/>
    <axis xyz="0 1 0"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <joint name="finger_upper_to_middle_joint_240" type="revolute">
    <parent link="finger_upper_link_240"/>
    <child link="finger_middle_link_240"/>
    <limit effort="1000" lower="0.0" upper="1.57" velocity="1000"/>
    <axis xyz="1 0 0"/>
    <origin rpy="0 0 0" xyz="0.01685 0.0505 0"/>
  </joint>
  <joint name="finger_middle_to_lower_joint_240" type="revolute">
    <parent link="finger_middle_link_240"/>
    <child link="finger_lower_link_240"/>
    <limit effort="1000" lower="-2.7" upper="0.0" velocity="1000"/>
    <axis xyz="1 0 0"/>
    <origin rpy="0 0 0" xyz="0.04922 0 -0.16"/>
  </joint>
  <joint name="holder_to_finger_0" type="fixed">
    <parent link="upper_holder_link"/>
    <child link="finger_base_link_0"/>
    <origin rpy="0 0 0.0" xyz="0 0 0"/>
  </joint>
  <joint name="holder_to_finger_120" type="fixed">
    <parent link="upper_holder_link"/>
    <child link="finger_base_link_120"/>
    <origin rpy="0 0 -2.0943951023931953" xyz="0 0 0"/>
  </joint>
  <joint name="holder_to_finger_240" type="fixed">
    <parent link="upper_holder_link"/>
    <child link="finger_base_link_240"/>
    <origin rpy="0 0 -4.1887902047863905" xyz="0 0 0"/>
  </joint>
</robot>
